### PR TITLE
Multithreading improvements

### PR DESCRIPTION
**Pull request by MerlinofMines**
Testing out a pull request by submitting the multi-threading changes as a pull request so pandincus can review

### CODE REVIEW

File: EasyCommands.Tests/ScriptTests/ScriptTest.cs [hunk 1]
```diff
@@ -89,6 +89,31 @@ public void RunUntilDone()
             RunUntil(t => Program.ProgramState.COMPLETE == Program.STATE || t.RunCounter >= 100);
         }
 
+        /// <summary>
+        /// Run the script a given number of iterations, regarldess of Program State.
```
Comment (MerlinofMines) [inline EasyCommands.Tests/ScriptTests/ScriptTest.cs:93]: typo

File: EasyCommands/CommandParsers/ParameterProcessors.cs [hunk 2]
```diff
@@ -916,26 +916,17 @@ public class WaitProcessor : SimpleCommandProcessor<WaitCommandParameter> {
             }
         }
 
-        public class ControlProcessor : SimpleCommandProcessor<ControlCommandParameter> {
```
Comment (MerlinofMines) [inline EasyCommands/CommandParsers/ParameterProcessors.cs:919]: I'm trying to get rid of SimpleCommandProcessor as I need to make changes to these processors, hence these changes

File: EasyCommands/Commands/Commands.cs [hunk 3]
```diff
@@ -20,20 +20,41 @@
...
+            bool Async;
+
+            public QueueCommand(Command command, bool Async) {
+                this.command = command;
+                this.Async = Async;
+            }
+
+            public override bool Execute() {
+                if(Async) {
+                    QueueAsyncThread(new Thread(command, "Queud", "Unknown"));
```
Comment (MerlinofMines) [inline EasyCommands/Commands/Commands.cs:50]: pretty sure that's now how Queued is spelled

File: EasyCommands/Commands/CommandParameters.cs [hunk 4]
```diff
@@ -75,6 +74,11 @@ public class ElseCommandParameter : CommandParameter { }
         public class PrintCommandParameter : CommandParameter { }
         public class SelfCommandParameter : CommandParameter { }
 
+        public class QueueCommandParameter : ValueCommandParameter<bool> {
+            public QueueCommandParameter(bool value) : base(value) {
```
Comment (MerlinofMines) [inline EasyCommands/Commands/CommandParameters.cs:78]: nit: could name this value better. it's intention is "IsAsync"

File: EasyCommands/Commands/Commands.cs [hunk 5]
```diff
@@ -107,53 +122,50 @@ public class VariableAssignmentCommand : Command {
...
             }
 
             public override bool Execute() {
                 switch (controlType) {
                     case ControlType.STOP:
-                        RUNNING_COMMANDS = null; STATE = ProgramState.STOPPED; break;
-                    case ControlType.PARSE:
-                        RUNNING_COMMANDS = null; ParseCommands(); STATE = ProgramState.STOPPED; break;
+                        THREAD_QUEUE.Clear();
+                        ASYNC_THREAD_QUEUE.Clear();
```
Comment (MerlinofMines) [inline EasyCommands/Commands/Commands.cs:144]: Should probably move these back to Program as some kinda "clear" function and make non-static so we can hide the queue logic inside Program only

File: EasyCommands/Program.cs [hunk 6]
```diff
@@ -25,15 +25,37 @@ public partial class Program : MyGridProgram {
         #region mdk preserve
         public static UpdateFrequency UPDATE_FREQUENCY = UpdateFrequency.Update1;
         public static LogLevel LOG_LEVEL = LogLevel.INFO;
-        public static int PARSE_AMOUNT = 1;
+        public static int FUNCTION_PARSE_AMOUNT = 1;
+        public static int MAX_ASYNC_THREADS = 50;
+        public static int MAX_QUEUED_THREADS = 50;
```
Comment (MerlinofMines) [inline EasyCommands/Program.cs:30]: I definitely picked these numbers out of my magic 8 ball

**Review by MerlinofMines — COMMENTED** [review]
K did a cursory review for myself.  Waiting for pandincus feedback

File: EasyCommands/Program.cs [hunk 7]
```diff
@@ -249,6 +310,23 @@ class CommandLine {
...
+            protected String name;
+
+            public Thread(Command command, string prefix, string name) {
+                this.command = command;
+                this.prefix = prefix;
+                this.name = name;
+            }
+
+            public Command GetCommand() => command;
+            public void SetCommand(Command c) => command = c;
```
Comment (pandincus) [inline EasyCommands/Program.cs:325]: nit: Replace with `public Command Command { get; set; }` on the declaration line.

**Review by pandincus — APPROVED** [review]
Looks awesome. Comments are almost entirely aesthetic in nature. SHIP IT.

File: EasyCommands.Tests/ScriptTests/MultiThreadingTests.cs [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+";
+
+            using (var test = new ScriptTest(script)) {
+                test.RunUntilDone();
+
+                Assert.AreEqual(2, test.Logger.Count);
+                Assert.AreEqual("Hello World", test.Logger[0]);
+                Assert.AreEqual("I Got Printed", test.Logger[1]);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (MerlinofMines) [reply EasyCommands.Tests/ScriptTests/MultiThreadingTests.cs:23]: You are 100% correct, will fix and verify

File: EasyCommands.Tests/ScriptTests/MultiThreadingTests.cs [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+
+                test.RunOnce();
+
+                Assert.AreEqual(2, test.Logger.Count);
+                Assert.AreEqual("I Got Printed", test.Logger[1]);
+            }
+        }
+
+        [TestMethod]
+        public void runArgumentIsQueudAtFront() {
```
[the remark this answers is not in the document]
Comment (MerlinofMines) [reply EasyCommands.Tests/ScriptTests/MultiThreadingTests.cs:48]: fixed

File: EasyCommands/Commands/Commands.cs [hunk 10]
```diff
@@ -20,20 +20,41 @@
...
-            }
+            public virtual Command Clone() { return this; }
 
             //Returns true if the program has finished execution.
             public abstract bool Execute();
         }
 
+        public class QueueCommand : Command {
+            public Command command;
+            bool Async;
```
[the remark this answers is not in the document]
Comment (MerlinofMines) [reply EasyCommands/Commands/Commands.cs:41]: fixed

File: EasyCommands/Commands/Commands.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (MerlinofMines) [reply EasyCommands/Commands/Commands.cs:50]: fixed

File: EasyCommands/Program.cs (hunk 7 above)
Comment (MerlinofMines) [reply EasyCommands/Program.cs:325]: fancy! fixed

**Review by pandincus — APPROVED** [review]
LGTM

**Pull request merged**